pyquil/tests/test_reference_density_simulator.py

```diff
@@ -255,3 +263,53 @@ def test_multiqubit_decay_bellstate():
     qam.execute(program)
 
     assert np.allclose(qam.wf_simulator.density, state)
+
+
+def test_for_negative_probabilities():
+    # trivial program to do state tomography on
+    prog = Program(I(0))
+
+    # make TomographyExperiment
+    expt_settings = [ExperimentSetting(zeros_state([0]), pt) for pt in [sI(0), sX(0), sY(0), sZ(0)]]
+    experiment_1q = TomographyExperiment(settings=expt_settings, program=prog)
+
+    # make an abstract compiler
+    class DummyCompiler(AbstractCompiler):
+        def get_version_info(self):
+            return {}
+
+        def quil_to_native_quil(self, program: Program):
+            return program
+
+        def native_quil_to_executable(self, nq_program: Program):
+            return nq_program
+
+    # make a quantum computer object
+    device = NxDevice(nx.complete_graph(1))
+    qc_density = QuantumComputer(name='testy!',
+                                 qam=PyQVM(n_qubits=1,
+                                           quantum_simulator_type=ReferenceDensitySimulator),
+                                 device=device,
+                                 compiler=DummyCompiler())
+
+    # initialize with a mixed state
+    initial_density = np.array([[1.0, 0.0], [0.0, 0.0]])
+    qc_density.qam.wf_simulator.density = initial_density
+
+    logger = logging.Logger('catch_all')
+
+    try:
+        list(measure_observables(qc=qc_density, tomo_experiment=experiment_1q, n_shots=3000))
+    except Exception as e:
+        print(logger.error('Error: ' + str(e)))
+        assert str(e) != 'probabilities are not non-negative'
+
+    # initialize with a mixed state
+    initial_density = np.array([[0.9, 0.0], [0.0, 0.1]])
+    qc_density.qam.wf_simulator.density = initial_density
+
+    try:
+        list(measure_observables(qc=qc_density, tomo_experiment=experiment_1q, n_shots=3000))
+    except Exception as e:
+        print(logger.error('Error: ' + str(e)))
+        assert str(e) != 'probabilities are not non-negative'
```

Comment:
I'm not sure I like the string comparison for error checking here. I think the more typical thing would be to define an exception `NegativeProbabilties`, and then catch this exception with something like
```
except NegativeProbabilities as e:
    pytest.fail(str(e))
```

Aside: "probabilities are not non-negative" -- couldn't this be the less-of-a-mouthful "probabilities are negative".

Reply:
- [x] todo

As you suggested offline, I added a `ValueError` exception rather than the generic exception